### ISSUE TITLE
The interface

### ISSUE THREAD

**Issue by iHiD** [exercism/automated-analysis#2]
I propose the following pattern for how this works:
- An orchestrating service ("Orchestrator") will download a solution into a directory ("Directory")
- Orchestrator will call a shell-script to run a language-specific static analyser ("Analyser"), passing the Directory as a paramater.
- Analyser will read the solution file(s) from the Directory, analyse them, write a JSON file called "analysis.json" with the results into the Director, and return a "success" exit code. If Analyser cannot achieve this, it will return a "failure" exit code.
- Orchestrator will respond to the script finishing and post information back to the website.

For now, I I propose the following beautifully simple JSON file structure: 
```json
{
  "approve": BOOLEAN
}
```

Your thoughts are welcome :)

**Comment by ErikSchierboom** [comment]
I like the shell script/directory based approach. The JSON format looks simple enough. I assume that comments are something that will be focused on later? That said, I think we'll also want to post a comment when approving an exercise: "great job" or something similar. Is this text something the analyzer should return or will that be done automatically?

**Comment by ErikSchierboom** [comment]
Addendum regarding the comments. There will be solutions that can be auto approved, but that will also really benefit from an additional comment. Something like: "Great work!... If you'd like to do some further iterating, you could try X..." This would require an additional comments property. We can of course also enabled this feature later on.

**Comment by tehsphinx** [comment]
### Current state of exercism

Exercism basically has 2 outputs currently:

1) A `response` struct from which a complete response to the student can be built:

```
type Response struct {
	// the following slices make up the answer in the order given here
	greeting    []gtpl.Template		// saying "Hi ..."
	intro       []gtpl.Template		// A little comment on their nice solution
	todo        []gtpl.Template		// Critical changes to be made before approval
	improvement []gtpl.Template		// Improvement suggestions
	comment     []gtpl.Template		// Thoughts and alternatives
	blocksugg   []gtpl.Template		// Block of information added as additional info, e.g. about benchmarking, comments in go, etc. Usually added as additional info on a suggestion.
	tip         []gtpl.Template		// Random tip on some valuable blog post, youtube talk or other resource
	outro       []gtpl.Template		// Something like: "If you have any questions..."
}
```

Note: Especially when auto-approving I very much like the feature of a **random tip** that provides extra food for the students. The random tip is usually a reference to a talk or a blog post. (Students regularly mention the tips as very valuable.)

2) An approving suggestion to the mentor

The ApproveSuggestion depends on the following information
```
type ApprSuggestion {
	gofmtPassed bool	// formatting ok?
	golintPassed bool	// linter happy?
	hasTodo bool		// any suggestion a todo? Todos are considered critical changes to be made before approving
	lenImprovements int	// too many improvement suggestions -> no approval
}
```

Note: The first 3 are blockers for approve (with exceptions on the first 2 exercises `two-fer` and `hamming`).

**Comment by tehsphinx** [comment]
**TLDR**: I agree with @ErikSchierboom that we should add a `response` field (string) to the json. I think it should be optional though so the hurdle for tools is small. If empty, no comment is added.

**Comment by ErikSchierboom** [comment]
I have a preference for using `comment` instead of `response`, as that better matches the terminology on the website I think. A sample payload would look like:

```json
{
  "approve": BOOLEAN,
  "comment": STRING (optional)
}
```

Feel free to shut this discussion down @iHiD, if you feel we're getting too far ahead of ourselves. It's just so much fun to think of how awesome this tool can be!

**Comment by tehsphinx** [comment]
Yeah, this really is exciting!

I wonder if the JSON should (maybe optionally) contain an additional "error" field in case of exit code != 0 (failure). That could give the calling routine more information on what went wrong. Mostly for debugging purposes.

Additionally we might want to define a few exit codes if e.g. a json file cannot be created. Suggestion:
```
0	Success
1	General Failure (optionally: more info can be written to json)
5	Disk Read Failure
6	Disk Write Failure
```

Feel free to add / remove / edit!

**Comment by cmccandless** [comment]
> I wonder if the JSON should (maybe optionally) contain an additional "error" field in case of exit code != 0 (failure)

Might I suggest optional field

```JSON
"error": {
    "code": 1,
    "message": [
        "First error output line",
        "..."
    ]
}
```

**Comment by ErikSchierboom** [comment]
I was wondering if instead of using a boolean property for approval, we could use an enumeration:

- `Approved`
- `RequiresChange`
- `RequiresMentoring`

The rationale being that compile errors or failing tests should result in an automatic comment prompting the user to change, and that these solutions do not need to be reviewed by a mentor.

**Comment by cmccandless** [comment]
So much for "beautifully simple" :joy:

**Comment by tehsphinx** [comment]
@cmccandless: About the error codes: I was only suggesting them as exit codes when a json file cannot be written.
Defining general error codes will probably be a never ending story as there will a huge variety of different errors that can happen in all those different languages. That would raise the bar of implementing the different tools considerably.

The error itself I would suggest as a basic string. New lines can be included as \n if needed.

**Comment by ErikSchierboom** [comment]
To be honest, I'm not really in favor of the error codes scheme. It seems a bit..complicated to me. Wouldn't some logging by the analyzer itself be enough to diagnose any potential issues?

This would be my proposed scheme, which I think it still quite simple (just to `string` fields):

```
{
  "approval": STRING [approved|requiresChange|requiresMentoring],
  "comment": STRING (optional)
}

**Comment by tehsphinx** [comment]
@ErikSchierboom 
So you are suggesting to write a log file to the directory for errors? Probably a good idea, I agree.

Still wonder if there should be some way of letting exercism know if there is a read/write disk failure. In that case a log file wouldn't help. That is where I thought it might help to define some (only very few!!) exit codes, as the tool should exit with an exit code anyway.

**Comment by iHiD** [comment]
Good discussion :)

I definitely think this JSON file will expand to have support for messages
etc!

However for a first cut of integration I'd like to keep the functionality
very simple - approved or not, as that's the MVP (minimal viable product
we're building) as a first stage.

I wouldn't conceptually mind another "status" key that recorded whether the
program ran successfully or not, and maybe the writing of a separate error
file if the process errors. However I consider the JSON file the response
that the UI will respond to (and probably content that we'll store directly
in the db). The UI (probably) won't care why or whether the automation
failed as we're not exposing that to the user (at least at this stage).

I guess, to sum up my thoughts, my focus is taking the smallest possible
steps right now. That doesn't mean the analysers can't be generating
messages internally etc, but until we've fleshed out how they work in terms
of the product, I'm wary to commit to how they fit into the analysers
output.

**Comment by ErikSchierboom** [comment]
> However for a first cut of integration I'd like to keep the functionality very simple - approved or not, as that's the MVP (minimal viable product we're building) as a first stage.

I understand the goal of building a MVP and keeping it very simple. Having just the approval status is fine of course. However, I do want to argue against using a boolean value for the approval status. In my experience, while using a boolean seems a very natural choice for a lot of things, in practice it can lead to some modelling issues later. The main problem with booleans is that they are, by definition, not extendable. If you later on decide that you want to add a third status, one cannot just modify the boolean value. Usually, what people then do is to add _another_ boolean flag, which can make things more complex as you now have four states to consider (true/true, true/false, false/true, false/false) instead of three. In my experience, this entire problem can be mitigated by using enumerations from the start.

Just my 2 cents :)

**Comment by iHiD** [comment]
Yeah. I totally agree about booleans in general. The reason I think that it is appropriate here is that the whole web-model is so tightly coupled to this. Approving a solution is the key trigger point in a solution's lifecycle. 

We will at some point definitely have a range of wider outputs. I'm imagining [approve without comment, approve with comment, disapprove with comment, disapprove and suggest comment to mentor, no idea, unable to parse] etc as the first few that come to mind (and you're right that there will be loads I've not considered)

Regardless of however many we have though, I want to know if we're approving this solution or not, and so would rather encode that as `{approve: true, messages: []}` or `{approve: true, messages: [...]}` than I would have a cartesian-set of status codes etc.

Thoughts?

**Comment by ErikSchierboom** [comment]
I understand. Let's leave it at this then and see how things will turn out later.

**Comment by tehsphinx** [comment]
I'd hoped for a commenting possibility right from the start but of course I understand that you have to figure out a lot of other things right now to get the entire infrastructure working.

Let's start with the MVP 😄

**Comment by iHiD** [comment]
Thanks @tehsphinx. We may well introduce commenting relatively early as "proposed comments for mentors" (similar to what exalysis does now) to get a feel for how comments are used by mentors. 

But (as you allude to) let's get it right rather than get there fast.

**Comment by ErikSchierboom** [comment]
Just wanted to mention it here. It might be wise to add a "version" field to the output file, to be able to evolve the output file with the least chance of breaking changes. This is likely not at all MVP material, but I just wanted to mention it.

**Comment by tehsphinx** [comment]
I just cannot shake the thought that the automated tools should be able to write comments also **completely autonomous**. Here some typical situations where I feel that is necessary:

(talking of "the tool" as a person 😄)
- A solution can basically be approved but there is a minor hint the tool would like to give the student. If the tool cannot do that there are 2 options: 1) leave it alone for a mentor to pick up (who might not make that comment after all) 2) approve without a comment which means loosing an opportunity.
- A solution is perfect but the tool still wants to give the student some additional studying material and some praise to encourage them.
- A solution needs some work and the tool knows what needs to be done.

Last point could also be done as a proposal to a mentor. But I am also thinking back on `rikki`: It had things to say about not so perfect solutions which helped get rid of the common mistakes. That would avoid additional work for mentors if solutions are already pre-mentored by the tool. It also gives some fast feedback to the students and avoids wait time on those iterations.

**Comment by ErikSchierboom** [comment]
Oh, and last thing about using booleans: they might introduce unwanted combinations. Suppose we have these two booleans: `approve` and `refer_to_mentor`. These values being booleans, users could conceivably return a status field in which both fields are `true`, which doesn't really make sense. Of course we could mention this in documentation and return an error when these values are passed, but the enumeration approach could help in prevent this.

**Comment by iHiD** [comment]
> I just cannot shake the thought that the automated tools should be able to write comments also completely autonomous.

I entirely agree with this and I agree with all the examples you give. But phase 1 is just approve/disapprove for our sanity! :)

**Comment by iHiD** [comment]
So @ErikSchierboom was right and I was wrong :)

I've pushed an interface doc [here](https://github.com/exercism/automated-mentoring-support/blob/master/docs/interface.md#output-format)

@ErikSchierboom @tehsphinx If you follow the 1/2/3 instructions [here](https://github.com/exercism/automated-mentoring-support/blob/master/docs/creating-an-analyzer.md) then I will do 5 for you.

**Comment by ErikSchierboom** [comment]
Really like the simple interface definition!

**Comment by tehsphinx** [comment]
@iHiD, @ErikSchierboom:  There is one more thing about the interface that I have been pondering a bit the last few days. We might need this at some point after we are able to make comments.

I think it would be good for an analyser to know on which iteration it is -- and maybe even what recommendations were already given.

Reason: We don't want to keep students in an endless loop repeating the same comments again and again.

Possible solution for the tricky part of remembering what recommendations were already given: Maybe the analysers could give a string to exercism to be kept for next iteration. That way they could self-manage keeping track of things.

Note: this is of course not for right now, but for later. Just writing it down somewhere.

**Comment by iHiD** [comment]
Yeah. I've thought a bit about this too and am unsure what the best approach is. I *think* we can say that the analyser should just send back it's comment(s) and if that's the same as a previous comment then the orchestrator overrides the analyzer and refers to a mentor (as we can presume the student hasn't understood the message). Needs lots of thought though.

FWIW, this is the reason I'm so keen to take this I'm babysteps across multiple different langs/analysers because we'll all have different ideas approaches, which we can consider at each step. If we try and work all this out upfront, it's going to either become very overwhelming or very top-down-led, neither of which we want.

Discussing here is perfect. But I'm keen to move slowly and really get this right to lay the groundwork work for the next 10m users so we don't end up in a world of hurt :)

(I know you agree with this too :))
